cmd/csi-external-health-monitor-controller/main.go

```diff
@@ -0,0 +1,303 @@
+		factory.Core().V1().PersistentVolumeClaims(), factory.Core().V1().Pods(), factory.Core().V1().Nodes())
+
+	run := func(ctx context.Context) {
+		stopCh := ctx.Done()
+		factory.Start(stopCh)
+		monitorController.Run(int(*workerThreads), stopCh)
+	}
+
+	if !*enableLeaderElection {
+		run(context.TODO())
+	} else {
+		// Name of config map with leader election lock
+		lockName := "external-health-monitor-leader-" + storageDriver
+		le := leaderelection.NewLeaderElection(clientset, lockName, run)
+
+		if *leaderElectionNamespace != "" {
+			le.WithNamespace(*leaderElectionNamespace)
+		}
+
+		if err := le.Run(); err != nil {
+			klog.Fatalf("failed to initialize leader election: %v", err)
+		}
+	}
+
+}
+
+func buildConfig(kubeconfig string) (*rest.Config, error) {
+	if kubeconfig != "" {
+		return clientcmd.BuildConfigFromFlags("", kubeconfig)
+	}
+	return rest.InClusterConfig()
+}
+
+// TODO: move this to csi-lib-utils
+func supportGetVolume(ctx context.Context, csiConn *grpc.ClientConn) (supportGetVolume bool, err error) {
+	client := csi.NewControllerClient(csiConn)
+	req := csi.ControllerGetCapabilitiesRequest{}
+	rsp, err := client.ControllerGetCapabilities(ctx, &req)
+	if err != nil {
+		return false, err
+	}
+
+	for _, cap := range rsp.GetCapabilities() {
+		if cap == nil {
+			continue
+		}
+		rpc := cap.GetRpc()
+		if rpc == nil {
+			continue
+		}
+		t := rpc.GetType()
+		if t == csi.ControllerServiceCapability_RPC_GET_VOLUME {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// TODO: move this to csi-lib-utils
+func supportGetVolumeCondition(ctx context.Context, csiConn *grpc.ClientConn) (supportGetVolumeHealth bool, err error) {
+	client := csi.NewControllerClient(csiConn)
+	req := csi.ControllerGetCapabilitiesRequest{}
+	rsp, err := client.ControllerGetCapabilities(ctx, &req)
+	if err != nil {
+		return false, err
+	}
+
+	for _, cap := range rsp.GetCapabilities() {
+		if cap == nil {
+			continue
+		}
+
+		rpc := cap.GetRpc()
+		if rpc == nil {
+			continue
+		}
+		t := rpc.GetType()
+		if t == csi.ControllerServiceCapability_RPC_VOLUME_CONDITION {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// TODO: move this to csi-lib-utils
+func supportsPluginControllerService(ctx context.Context, csiConn *grpc.ClientConn) (bool, error) {
+	client := csi.NewIdentityClient(csiConn)
+	req := csi.GetPluginCapabilitiesRequest{}
+	rsp, err := client.GetPluginCapabilities(ctx, &req)
+	if err != nil {
+		return false, err
+	}
+	for _, cap := range rsp.GetCapabilities() {
+		if cap == nil {
+			continue
+		}
+		srv := cap.GetService()
+		if srv == nil {
+			continue
+		}
+		t := srv.GetType()
+		if t == csi.PluginCapability_Service_CONTROLLER_SERVICE {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// TODO: move this to csi-lib-utils
+// TODO: do we need to consider ControllerServiceCapability_RPC_LIST_VOLUMES_VOLUME_HEALTH ?
```

Comment:
oh, I added a comment above that LIST_VOLUMES should be checked and it should be the preferred way to retrieve information in batch.

Also you may be using an earlier version of the spec?  LIST_VOLUMES_VOLUME_HEALTH is removed.  Now we just have one VOLUME_CONDITION which is used for both LIST_VOLUMES and GET_VOLUME.